src/shape.cpp

```diff
@@ -490,6 +490,26 @@ shape shape::with_lens(const std::vector<std::size_t>& l) const
     return this->with_lens(this->type(), l);
 }
 
+shape shape::prepend_dim(size_t dim) const
+{
+    if(this->dynamic())
+        MIGRAPHX_THROW("SHAPE: with_lens() called on dynamic shape");
+    auto cur_lens = this->lens();
+    std::vector<std::size_t> nlens(1 + cur_lens.size(), dim);
+    std::copy(cur_lens.begin(), cur_lens.end(), nlens.begin() + 1);
+    return {this->type(), nlens};
+}
+
+shape shape::append_dim(size_t dim) const
+{
+    if(this->dynamic())
+        MIGRAPHX_THROW("SHAPE: with_lens() called on dynamic shape");
+    auto cur_lens = this->lens();
+    std::vector<std::size_t> nlens(1 + cur_lens.size(), dim);
+    std::copy(cur_lens.begin(), cur_lens.end(), nlens.begin());
+    return {this->type(), nlens};
```

Comment:
This completely clears the layout of the original shape. I am not sure if its good to add methods like this.

Reply:
yeah. We can probably add an assert/throw that it is a standard shape or better would be to remove it. I don't foresee this being used otherwise.

Reply:
We don't have a choice of clearing out the original shape.. imagine doing a prepend of a dimension with some expectations of its strides and whatnot..

The reason to do this api here is that it shouldn't be done in the actual operator code -- it belongs to shape.* logic -- rather than externalize it.

Thanks.

Reply:
> We don't have a choice of clearing out the original shape..

Which is why it makes this method problematic. This could inadvertently change the layout of the tensor, which can lead to bugs or errors in our code, especially when there are more correct ways to do this. For example, prepending or appending ones will lose the strides, whereas using the equivalent squeeze and unsqueeze will not. 

If you dont care about the strides, then you dont need the shape class at all, you just need a vector of dimensions. And in your case you never look at the strides, which means you dont need the shape class.

Reply:
@pfultz2 , do you have a solution for this? Objections, I understand.

I want a method that prepends the lens, and a method that appends a shape. Do you want me to do it manually in that parse_operator codebase. I will be happy to do it. Thanks.

Reply:
removed from shape.cpp/hpp.